docs/src/main/asciidoc/writing-extensions.adoc

```diff
@@ -612,6 +612,8 @@ This is a very flexible mechanism.
 NOTE: `BuildItem` instances should be immutable, as the producer/consumer model does not allow for mutation to be correctly
 ordered. This is not enforced but failure to adhere to this rule can result in race conditions.
 
+NOTE: Build steps are only executed if and only if they produce build items that are (transitively) needed by other build steps. Make sure your build step produces a build item, otherwise you should probably produce either `ValidationErrorBuildItem` for build validations, or `ArtifactResultBuildItem` for generated artifacts.
```

Comment:
```suggestion
NOTE: Build steps are executed if and only if they produce build items that are (transitively) needed by other build steps. Make sure your build step produces a build item, otherwise you should probably produce either `ValidationErrorBuildItem` for build validations, or `ArtifactResultBuildItem` for generated artifacts.
```